### PR TITLE
Feature/fix large codelist filter timeout

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

Improved performance of HierarchyUpdate calls:
- Before: 
  - `RemoveDimensionValue` called for each dimension option that needs to be removed
  - `AddDimensionValue` called for each dimension option that needs to be added
- Now:
  - `GetDimensionOptions` called once
  - Do the changes in the array of options locally
  -  `AddDimensionValues` called once with the modified array of options

Note that AddDimensionValues will override the whole dimension options array, that's why we get the original values, modify them and send the modified version.

A better solution would be to modify the filter API spec, adding a `PATCH /filters/{filter_blueprint_id}/dimensions/{name}` endpoint. I described this spike here:
https://docs.google.com/document/d/16kEjaFanTXtNdSAvJF8ux6iUdrGCY8cb9VguLa_46T8/edit#

In any case, if we decide to implement the PATCH operation, the change implemented in this PR will be useful, but will need to be modified to create the patch instead of modifying the full array of items.

### How to review

- Make sure code changes make sense and improve performance.
- Unit tests should pass

### Who can review

Anyone.

### CODE REVIEW

File: handlers/hierarchy.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (f *Filter) HierarchyUpdate(w http.ResponseWriter, req *http.Request) {
 		}
 	}
 
-	fil, err := f.FilterClient.GetJobState(req.Context(), userAccessToken, "", collectionID, "", filterID)
+	fil, err := f.FilterClient.GetJobState(req.Context(), userAccessToken, "", "", collectionID, filterID)
```
Comment (DavidSubiros) [inline handlers/hierarchy.go:61]: I think this was a bug in the original code (we were switching collectionID and downloadServiceToken parameters according to GetJobState method signature in dp-api-clients-go)

Comment (nshumoogum) [reply handlers/hierarchy.go:60]: Probably was fine until we update the package here but yes I remember doing this as this seemed a more logical order

File: handlers/hierarchy.go [hunk 2]
```diff
@@ -135,13 +135,25 @@ func (f *Filter) HierarchyUpdate(w http.ResponseWriter, req *http.Request) {
...
-			log.Event(ctx, "failed to add dimension value", log.Error(err))
-		}
+		mapUniqueOptions[k] = struct{}{}
 	}
 
-	wg.Wait()
-	http.Redirect(w, req, redirectURI, 302)
+	// Efficiently generate slice from map keys:
+	patchedOptionsValues := make([]string, len(mapUniqueOptions))
+	i := 0
```
Comment (nshumoogum) [inline handlers/hierarchy.go:143]: Is it necessary to use a counter? Can we not use append function to `patchedOptionsValues`?

The other thing might be to handle if the counter exceeds `len(mapUniqueOptions)`  to throw an error and return internal server error; then we would want to keep the count variable in `i`

File: main.go [hunk 3]
```diff
@@ -71,7 +79,7 @@ func main() {
 	clients.Healthcheck = &healthcheck
 
 	if err = registerCheckers(ctx, clients); err != nil {
-		os.Exit(1)
+		return err
 	}
```
Comment (nshumoogum) [inline main.go:84]: Line 100 - Should return error

File: handlers/hierarchy.go (hunk 2 above)
Comment (DavidSubiros) [reply handlers/hierarchy.go:143]: I did it this way to increase efficiency, but I think it's not necessary - I will use `append`

File: main.go (hunk 3 above)
Comment (DavidSubiros) [reply main.go:84]: Shall I create an error channel and log it?

Comment (DavidSubiros) [reply main.go:84]: But it's a separate go routine. I think in some services we had a channel, and then if it was closed, we would start the graceful shutdown - but after the requirement to not kill services if dependencies fail, the channel didn't have any propose (we log anyway from the go routine if an error happens)

Comment (nshumoogum) [reply main.go:84]: I think we should be forcing a graceful shutdown here as this application has only one responsibility and that is to handle requests. If the server is an issue then the application can't function. I would force a graceful shutdown here

**Pull request merged**